piptools/utils.py

```diff
@@ -152,89 +169,33 @@ def flat_map(fn, collection):
     return chain.from_iterable(map(fn, collection))
 
 
-def lookup_table(values, key=None, keyval=None, unique=False, use_lists=False):
+def lookup_table(
+    values: Iterable[Union[_VT, Tuple[_KT, _VT]]],
+    key: Optional[Callable[[_VT], _KT]] = None,
+) -> Dict[_KT, Set[_VT]]:
     """
     Builds a dict-based lookup table (index) elegantly.
-
-    Supports building normal and unique lookup tables.  For example:
-
-    >>> assert lookup_table(
-    ...     ['foo', 'bar', 'baz', 'qux', 'quux'], lambda s: s[0]) == {
-    ...     'b': {'bar', 'baz'},
-    ...     'f': {'foo'},
-    ...     'q': {'quux', 'qux'}
-    ... }
-
-    For key functions that uniquely identify values, set unique=True:
-
-    >>> assert lookup_table(
-    ...     ['foo', 'bar', 'baz', 'qux', 'quux'], lambda s: s[0],
-    ...     unique=True) == {
-    ...     'b': 'baz',
-    ...     'f': 'foo',
-    ...     'q': 'quux'
-    ... }
-
-    For the values represented as lists, set use_lists=True:
-
-    >>> assert lookup_table(
-    ...     ['foo', 'bar', 'baz', 'qux', 'quux'], lambda s: s[0],
-    ...     use_lists=True) == {
-    ...     'b': ['bar', 'baz'],
-    ...     'f': ['foo'],
-    ...     'q': ['qux', 'quux']
-    ... }
-
-    The values of the resulting lookup table will be lists, not sets.
-
-    For extra power, you can even change the values while building up the LUT.
-    To do so, use the `keyval` function instead of the `key` arg:
-
-    >>> assert lookup_table(
-    ...     ['foo', 'bar', 'baz', 'qux', 'quux'],
-    ...     keyval=lambda s: (s[0], s[1:])) == {
-    ...     'b': {'ar', 'az'},
-    ...     'f': {'oo'},
-    ...     'q': {'uux', 'ux'}
-    ... }
-
     """
-    if keyval is None:
-        if key is None:
 
-            def keyval(v):
-                return v
+    def keyval(v: Union[_VT, Tuple[_KT, _VT]]) -> Tuple[_KT, _VT]:
+        if isinstance(v, tuple):
+            return v[0], v[1]
 
-        else:
+        assert key is not None, "key function must be specified"
```

Comment:
This is a serious misuse of the `assert` statement. It must only check things that should always be true in runtime but here it seems to be used as an input validation. Moreover, it's an input validation of another function (`lookup_table()`).
`asserts` are stripped off in the optimized mode (except for the pytest runtime because it patches the bytecode so this would be different from the real runtime).
Another problem is that it's a check that (when enabled) is made multiple times, on each invocation in the loop. It should instead be done in the `lookup_table()` function itself.
Besides, this sort of checks should be done with an `if`-statement and a condition, followed by raising an appropriate exception explicitly. Raising an `AssertionError` is not a good API for this. An appropriate error to raise would be a `TypeError` or a `ValueError`.
```suggestion
```

Reply:
Good point, thanks. Addressed in 1c125852462698c1d40e50f796f81c73f3411f2b.

Reply:
Note, that `assert key is not None` guard is still there to satisfy `mypy`. Any ideas to improve this?

Reply:
`typing.cast()`?

Reply:
Could work, but doesn't look like an improvement. Maybe leave as is?

Reply:
ok